modules/navigation/3d/nav_mesh_queries_3d.cpp

```diff
@@ -127,87 +128,222 @@ Vector3 NavMeshQueries3D::polygons_get_random_point(const LocalVector<gd::Polygo
+		case NavigationPathQueryParameters3D::PathPostProcessing::PATH_POSTPROCESSING_CORRIDORFUNNEL: {
+			query_task.path_postprocessing = PathPostProcessing::PATH_POSTPROCESSING_CORRIDORFUNNEL;
+		} break;
+		case NavigationPathQueryParameters3D::PathPostProcessing::PATH_POSTPROCESSING_EDGECENTERED: {
+			query_task.path_postprocessing = PathPostProcessing::PATH_POSTPROCESSING_EDGECENTERED;
+		} break;
+		case NavigationPathQueryParameters3D::PathPostProcessing::PATH_POSTPROCESSING_NONE: {
+			query_task.path_postprocessing = PathPostProcessing::PATH_POSTPROCESSING_NONE;
+		} break;
+		default: {
+			WARN_PRINT("No match for used PathPostProcessing - fallback to default");
+			query_task.path_postprocessing = PathPostProcessing::PATH_POSTPROCESSING_CORRIDORFUNNEL;
+		} break;
+	}
+
+	query_task.metadata_flags = (int64_t)p_query_parameters->get_metadata_flags();
+	query_task.simplify_path = p_query_parameters->get_simplify_path();
+	query_task.simplify_epsilon = p_query_parameters->get_simplify_epsilon();
+	query_task.status = NavMeshPathQueryTask3D::TaskStatus::QUERY_STARTED;
+
+	map->query_path(query_task);
+
+	const uint32_t path_point_size = query_task.path_points.size();
+
+	Vector<Vector3> path_points;
+	Vector<int32_t> path_meta_point_types;
+	TypedArray<RID> path_meta_point_rids;
+	Vector<int64_t> path_meta_point_owners;
+
+	{
+		path_points.resize(path_point_size);
+		Vector3 *w = path_points.ptrw();
+		const Vector3 *r = query_task.path_points.ptr();
+		for (uint32_t i = 0; i < path_point_size; i++) {
+			w[i] = r[i];
 		}
+	}
 
-		// For each face check the distance between the origin/destination
-		for (size_t point_id = 2; point_id < p.points.size(); point_id++) {
-			const Face3 face(p.points[0].pos, p.points[point_id - 1].pos, p.points[point_id].pos);
+	if (query_task.metadata_flags.has_flag(PathMetadataFlags::PATH_INCLUDE_TYPES)) {
+		path_meta_point_types.resize(path_point_size);
+		int32_t *w = path_meta_point_types.ptrw();
+		const int32_t *r = query_task.path_meta_point_types.ptr();
+		for (uint32_t i = 0; i < path_point_size; i++) {
+			w[i] = r[i];
+		}
+	}
+	if (query_task.metadata_flags.has_flag(PathMetadataFlags::PATH_INCLUDE_RIDS)) {
+		path_meta_point_rids.resize(path_point_size);
+		for (uint32_t i = 0; i < path_point_size; i++) {
+			path_meta_point_rids[i] = query_task.path_meta_point_rids[i];
+		}
+	}
+	if (query_task.metadata_flags.has_flag(PathMetadataFlags::PATH_INCLUDE_OWNERS)) {
+		path_meta_point_owners.resize(path_point_size);
+		int64_t *w = path_meta_point_owners.ptrw();
+		const int64_t *r = query_task.path_meta_point_owners.ptr();
+		for (uint32_t i = 0; i < path_point_size; i++) {
+			w[i] = r[i];
+		}
+	}
 
-			Vector3 point = face.get_closest_point_to(p_origin);
-			real_t distance_to_point = point.distance_to(p_origin);
-			if (distance_to_point < begin_d) {
-				begin_d = distance_to_point;
-				begin_poly = &p;
-				begin_point = point;
-			}
+	p_query_result->set_path(path_points);
```

Comment:
Most of the copies here are intentional as this is the place to step from user land to server internals.

The RefCounted Parameters and Result are user exposed. They can be broken at any time by users that do funky stuff with them.

This is why we immediatly copy from the parameters and later copy back to the result object not using anything from them in the actual server internals.

This will be even more important with threading where we can trust users even less to not break things. The system can be seen in the navmesh baking that uses similar logic but already uses the task for threaded bakes.